stdlib/asm/collections/smt.masm

```diff
@@ -0,0 +1,163 @@
+    # [N, d, K, R, V, ...]
+    adv_push.1
+    if.true
+        # make a domain word for depth
+        # [0, 0, d, 0, K, R, ...]
+        push.0 swap push.0.0
+
+        # load the value
+        # [V, 0, 0, d, 0, K, R, V, ...]
+        padw adv_loadw movdnw.3 dupw.3
+
+        # compute remaining path key
+        # [K', V, 0, 0, d, 0, K, R, V, ...]
+        dupw.2 dup.10 eq.16
+        if.true
+            exec.shr_16
+        else
+            dup.10 eq.32
+            if.true
+                exec.shr_32
+            else
+                dup.10 eq.48
+                if.true
+                    exec.shr_48
+                else
+                    dup.10 eq.64
+                    if.true
+                        # remaining path of depth 64 is 0
+                        drop push.0
+                    else
+                        # adv provider returned an invalid depth
+                        # panic
+                        push.0
+                        assert
+                    end
+                end
+            end
+        end
+
+        # perform the permutation and keep only digest output
+        # [N, d, K, R, V, ...]
+        swapw dup.10 movdn.12 hperm dropw swapw dropw
+    else
+        # [d, K, R, V, ...]
+        padw padw drop movdnw.3 push.0 dropw
+
+        # push empty constant as node
+        # [N, d, K, R, V, ...]
+        dup eq.16
+        if.true
+            push.17483286922353768131
+            push.353378057542380712
+            push.1935183237414585408
+            push.4820339620987989650
+        else
+            dup eq.32
+            if.true
+                push.11677748883385181208
+                push.15891398395707500576
+                push.3790704659934033620
+                push.2126099371106695189
+            else
+                dup eq.48
+                if.true
+                    push.10650694022550988030
+                    push.5634734408638476525
+                    push.9233115969432897632
+                    push.1437907447409278328
+                else
+                    dup eq.64
+                    if.true
+                        padw
+                    else
+                        # adv provider returned an invalid depth
+                        # panic
+                        push.0
+                        assert
+                    end
+                end
+            end
+        end
+    end
+
+    # compute index
+    # [i, d, N, K, R, V, ...]
+    movup.4 dup eq.16
+    if.true
+        dup.5 exec.shr_48
+    else
+        dup eq.32
+        if.true
+            dup.5 exec.shr_32
+        else
+            dup eq.48
+            if.true
+                dup.5 exec.shr_16
+            else
+                # depth already checked; can assume 64
+                dup.5
+            end
+        end
+    end
+
+    # check if adv provider opens to root
+    # [N', R, N, K, V, ...]
+    push.0.0 movupw.3 swapw drop drop swap mtree_get
```

Comment:
I would move this up. Basically, I think it would be simpler if we do the following:

1. Compute index/remainder key.
2. Get the node from the tree using `mtree_get`.
3. Convert node to value (i.e., either extract value from node, or convert empty subtree root to zeros).

Reply:
The implementation of this function changed entirely. I'm resolving this discussion as it is no longer applicable